src/services/vyos-conntrack-logger

```diff
@@ -0,0 +1,365 @@
+    if tmp := conf_proto.get(SUPPORTED_PROTO_TO_NAME.get(proto_num, 'other')):
+        if tmp.get(event_type) is not None:
+            return True
+    return False
+
+
+def parse_conntrack_event(msg: nfct_msg, conf_proto: Dict) -> Dict:
+    data = dict()
+    event_type = parse_event_type(msg['header'])
+    proto_num = msg.get_nested('CTA_TUPLE_ORIG', 'CTA_TUPLE_PROTO', 'CTA_PROTO_NUM')
+
+    if not is_need_to_log(event_type, proto_num, conf_proto):
+        return data
+
+    data = {
+        'COMMON': {
+            'EVENT_TYPE': event_type,
+            'TIME_OUT': msg.get_attr('CTA_TIMEOUT'),
+            'MARK': msg.get_attr('CTA_MARK'),
+            'PORTID': msg['header'].get('pid'),
+            'PROTO_INFO': parse_proto_info(msg.get_attr('CTA_PROTOINFO')),
+            'STATUS': msg.get_attr('CTA_STATUS'),
+        },
+        'ORIG': {},
+        'REPLY': {},
+    }
+
+    for direct in ['ORIG', 'REPLY']:
+        data[direct]['ADDR'] = parse_ip_addr(msg['nfgen_family'], msg.get_attr(f'CTA_TUPLE_{direct}'))
+        data[direct]['PROTO'] = parse_proto(msg.get_attr(f'CTA_TUPLE_{direct}'))
+        data[direct]['COUNTERS'] = parse_counters(msg.get_attr(f'CTA_COUNTERS_{direct}'))
+
+    return data
+
+
+def get_config_event_groups(config: Dict) -> Set:
+    event_groups = set()
+
+    for key, val in config['protocols'].items():
+        event_groups.update(list(val.keys()))
+
+    return event_groups
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-c',
+                        '--config',
+                        action='store',
+                        help='Path to vyos-conntrack-logger configuration',
+                        required=True,
+                        type=Path)
+
+    args = parser.parse_args()
+    try:
+        config_path = Path(args.config)
+        config = json.loads(config_path.read_text())
+    except Exception as err:
+        print(
+            f'Configuration file "{config_path}" does not exist or malformed: {err}'
+        )
+        exit(1)
+
+    cfg_group = grp.getgrnam(CFG_GROUP)
+    os.setgid(cfg_group.gr_gid)
+
+    signal.signal(signal.SIGHUP, sig_handler)
+    signal.signal(signal.SIGTERM, sig_handler)
+
+    event_groups = get_config_event_groups(config)
+    conf_proto = config['protocols']
+
+    ct = conntrack.Conntrack()
+    ct.bind()
+
+    for name in event_groups:
+        if group := EVENT_NAME_TO_GROUP.get(name):
+            ct.add_membership(group)
+        else:
+            logger.error(f'Unexpected event group {name}')
+    try:
+        while True:
+            logger.debug('Enter main loop...')
+            if shutdown:
+                break
+            for msg in ct.get():
+                parsed_event = parse_conntrack_event(msg, conf_proto)
+
+                if parsed_event:
+                    message = format_event_message(parsed_event)
+                    logger.info(message)
```

Comment:
I think we need to use queues with threads here to keep control over the daemon in high loads and utilize all available CPU cores for processing.

Reply:
done.
One thread receives socket messages and puts them into the queue, several processes (one per CPU) get messages from the queue and parse them, write log.

NOTE: In the current implementation, the order of messages in the log may be disrupted because messages are parsed asynchronously

Reply:
> NOTE: In the current implementation, the order of messages in the log may be disrupted because messages are parsed asynchronously

This is a problem. A common use case for this feature is to detect a connection's start and end times.